File: src/main/java/com/wikia/tibia/repositories/WikiObjectRepository.java
package com.wikia.tibia.repositories;

import com.wikia.tibia.gateways.WikiObjectGateway;
import com.wikia.tibia.jackson.Parser;
import com.wikia.tibia.objects.WikiObject;
import io.vavr.control.Try;

import java.util.List;

public abstract class WikiObjectRepository<T extends Class<? extends WikiObject>, U extends WikiObjectGateway<WikiObject>> {

    private Class<T> wikiObjectClass;
    private U wikiObjectGateway;

    public WikiObjectRepository(Class<T> wikiObjectClass, U wikiObjectGateway) {
        this.wikiObjectClass = wikiObjectClass;
        this.wikiObjectGateway = wikiObjectGateway;
    }

    public Try<List<T>> getWikiObjects() {
        return wikiObjectGateway.getWikiObjects(true)
                .map(json -> Parser.listOneByOne(wikiObjectClass, json, -1));
    }

    public Try<List<T>> getWikiObjects(long limit) {
        return wikiObjectGateway.getWikiObjects(true)
                .map(json -> Parser.listOneByOne(wikiObjectClass, json, limit));
    }

    public Try<List<String>> getWikiObjectsList() {
        return wikiObjectGateway.getWikiObjects(false)
                .map(json -> Parser.list(String.class, json));
    }

    public Try<String> saveWikiObject(WikiObject wikiObject, String editSummary, boolean dryRun) {
        return wikiObjectGateway.saveWikiObject(wikiObject, editSummary, dryRun);
    }

    public Try<Object> getWikiObject(String pageName) {
        return wikiObjectGateway.getWikiObject(pageName)
                .map(json -> Parser.parse(wikiObjectClass, json));
    }
}

File: src/main/kotlin/com/wikia/tibia/repositories/WikiObjectRepository.kt
package com.wikia.tibia.repositories

import com.wikia.tibia.gateways.WikiObjectGateway
import com.wikia.tibia.jackson.Parser.list
import com.wikia.tibia.jackson.Parser.listOneByOne
import com.wikia.tibia.jackson.Parser.parse
import com.wikia.tibia.objects.WikiObject
import io.vavr.control.Try

abstract class WikiObjectRepository<T : WikiObject>(
        private val wikiObjectClass: Class<T>,
        private val wikiObjectGateway: WikiObjectGateway<in WikiObject>
) {

    fun getWikiObjects(): Try<List<T>> {
        return wikiObjectGateway.getWikiObjects(true)
                .map { listOneByOne(type = wikiObjectClass, json = it) }
    }

    fun getWikiObjects(limit: Int = -1): Try<List<T>> {
        return wikiObjectGateway.getWikiObjects(true)
                .map { listOneByOne(type = wikiObjectClass, json = it, limit = limit) }
    }

    fun getWikiObjectsList(): Try<List<String>> {
        return wikiObjectGateway.getWikiObjects(false)
                .map { list(String::class.java, it) }
    }

    fun saveWikiObject(wikiObject: WikiObject, editSummary: String?, dryRun: Boolean): Try<String> {
        return wikiObjectGateway.saveWikiObject(wikiObject, editSummary, dryRun)
    }

    fun getWikiObject(pageName: String): Try<Any?> {
        return wikiObjectGateway.getWikiObject(pageName)
                .map { parse(wikiObjectClass, it) }
    }

}